test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +280,595 @@ func TestAdoption(t *testing.T) {
-
-	// change the rs's selector to match both pods
-	patch := `{"spec":{"selector":{"matchLabels": {"uniqueKey":"1"}},"template":{"metadata":{"labels":{"uniqueKey":"1"}}}}}`
-	rsClient := clientSet.Extensions().ReplicaSets(ns.Name)
-	rs, err := rsClient.Patch(rs.Name, types.StrategicMergePatchType, []byte(patch))
+	defer close(stopCh)
+	waitRSStable(t, c, rs, ns.Name)
+
+	testStatusReplicas(t, c, rs, ns.Name, 3)
+	testStatusReplicas(t, c, rs, ns.Name, 0)
+	testStatusReplicas(t, c, rs, ns.Name, 2)
+
+	// Update 2 pods that RS created such that one is failed and one is being deleted
+	// Ensure both are being created again in the end
+	podClient := c.Core().Pods(ns.Name)
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := podClient.List(options)
 	if err != nil {
-		t.Fatalf("Failed to patch replica set: %v", err)
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
 	}
-	t.Logf("patched rs = %#v", rs)
-	// wait for the rs to create one more pod
-	if err := wait.Poll(10*time.Second, 60*time.Second, func() (bool, error) {
-		return verifyRemainingObjects(t, clientSet, ns.Name, 1, 3)
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	if len(pods.Items) != 2 {
+		t.Fatalf("len(pods) = %v, want %v", len(pods.Items), 2)
+	}
+
+	failedPod := &pods.Items[0]
+	deletingPod := &pods.Items[1]
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(failedPod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		newPod.Status.Phase = v1.PodFailed
+		_, err = podClient.Update(newPod)
+		return err
 	}); err != nil {
-		t.Fatal(err)
+		t.Fatalf("Failed to set .Status.Phase of pod %s: %v", failedPod.Name, err)
 	}
-	podClient := clientSet.Core().Pods(ns.Name)
-	pod2, err = podClient.Get(pod2.Name, metav1.GetOptions{})
-	if err != nil {
-		t.Fatalf("Failed to get pod2: %v", err)
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(deletingPod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		return c.Core().Pods(ns.Name).Delete(newPod.Name, &metav1.DeleteOptions{})
+	}); err != nil {
+		t.Fatalf("Failed to delete pod %s: %v", deletingPod.Name, err)
 	}
-	if len(pod2.OwnerReferences) != 0 {
-		t.Fatalf("ownerReferences of pod2 is not cleared, got %#v", pod2.OwnerReferences)
+
+	rsClient := c.Extensions().ReplicaSets(ns.Name)
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if newRS.Status.Replicas != 2 {
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		t.Fatalf("Failed to verify .Status.Replicas is equal to .Spec.Replicas of rs %s: %v", rs.Name, err)
+	}
+
+	// Update .Generation and verify .Status.ObservedGeneration is changed
```

Comment:
Since this is a real API server, you can't directly modify Generation. You have to trigger the API server to do it for you by updating something in Spec.

The goal is to test that ObservedGeneration gets incremented to match Generation, even if nothing else about RS.Status needs to change. You can do that like this:

* Update something in RS Spec that doesn't cause it to touch Pods. I think you could add an annotation to the Pod template; RS doesn't do rolling update so it shouldn't care. Make sure not to change Pod template labels though, because RS will need to update status to reflect that.
* Store the RS that gets returned from Update(). This will contain the latest Generation value.
* Poll the RS until Status.ObservedGeneration >= this value.

Reply:
Agreed. Adding the annotation is the correct way for testing this. `Generation` is a field generated by the API server when the spec changes. Modifying `Generation` directly does not test anything at all.

Reply:
Use ObservedGeneration in general to ensure that ReplicaSetController has observed your modifications to the RS. This is sufficient to ensure that Generation/ObservedGeneration tracking works correctly.

Reply:
Noted, thanks for the details!

Reply:
@kow3ns This test is specifically to make sure ObservedGeneration gets updated even if nothing else in Status would otherwise change. Checking ObservedGeneration elsewhere is a good idea, but would be insufficient to ensure we produce a Status update in which nothing else changes.